### DISCUSSION TITLE
the idea of a more customizable skin

### DISCUSSION THREAD

**Discussion by nana123126** [ppy/osu#24847] [Ideas] [answered]
I noticed that the skin freedom of 'osu lazer' has greatly exceeded that of 'osu stable' (in the game), so--

I have an idea to make it completely customizable. It sounds a bit crazy, can skin elements be manipulated by programming languages? For example, using C# or Python to control the state and position of elements can make the skin more customizable.

It may be risky to use these high-level languages, which allows malicious code to run, so a better approach is to create a specialized language or use graphical programming.

Perhaps this may lead to cheating skins, such as avoiding the "FlashLight" mode by displaying the next circle position on the screen (if it is free enough to do so), and it may be necessary to disable the interface for reading the circle position, but this will also make it limited. I mean, skins that use these cheating level features can be classified as non rank, preventing players from obtaining PPs and leaderboards after playing, and marking their scores as untrusted. This is better than simply disabling the feature; This type of skin is suitable for players who want to relax.

--Other--

This looks like it will change a lot, and I think if we add skin interaction on top of it, does it mean that players can write UI now. It may run counter to the original intention of the skin, and simply changing the display content is more suitable for the concept of 'skin', as the current 'osu lazer' seems to be developing in this direction. Adding these features just for the sake of 'skin' may be too bulky.

It may be more suitable to directly obtain the code for 'osu lazer' and add features, but this is not rankable and unsupported. Instead, a new programming language is needed to solve this problem, and it can also make writing easier (to be honest, I have not studied C# because these high-level languages are all huge, and if you only create a programming language for skin systems, don't need many features, like Lua).

**Comment by peppy** [comment, marked as answer]
This has been discussed multiple times and is already basically supported, using building blocks (to be made). You won't be able to use c#.

Expect this in a distant future as we have other priorities.
